### DISCUSSION TITLE
Ozone key data structure.

### DISCUSSION THREAD

**Discussion by acedesai** [apache/ozone#6639] [General]
Hey guys, I was wondering whether Ozone can persist a specific set of metadata with a key (an object). This is similar to the awss3 specifications. If so, is there a reference for this exact data structure and api please?

**Comment by ivandika3** [comment]
Hi @acedesai , Ozone supports user-defined object metadata similar to AWS S3 specification since 1.3.0. See https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html for the AWS S3 user guide.

That being said, there were some recent major improvements and fixes regarding custom metadata:
- #6273
- #6490
- #6489

These improvements are planned to be included in the ongoing Ozone 1.4.1 release which also includes other S3 improvements and fixes. There is currently no target release date for 1.4.1 release (although it should be done in the very near future). So, you can choose to backport these improvements to 1.4.0 (latest version) manually or you can wait for the 1.4.1 release to be done.

For the data structure, `OmKeyInfo` contains a metadata field to support custom metadata. You can read the implementation deeply and let us know if you have any questions.

Hope this helps.
